### PR TITLE
feat: copy/paste step

### LINKED ISSUE #7: [FR] Copy/Paste step feature
Add the ability to copy and paste a step while in step edit mode.
PR incoming!

### PR DESCRIPTION
**Pull request by ajboni**
Default shortcuts ctr+c + ctrl+v
quit default shortcut is changed to ctrl+q

resolves #7

### CODE REVIEW

File: filesystem/keymap.go [hunk 1]
```diff
@@ -18,6 +18,8 @@ type KeyMap struct {
 	TempoDown    string     `json:"tempo_down"`
 	AddParam     string     `json:"add_param"`
 	RemoveParam  string     `json:"remove_param"`
+	CopyParams   string     `json:"copy_params"`
+	PasteParams  string     `json:"paste_params"`
```
Comment (xaviergodart) [inline filesystem/keymap.go]: I think `CopyStep` and `PasteStep` would be more appropriate here.

File: sequencer/sequencer.go [hunk 2]
```diff
@@ -245,6 +249,69 @@ func (s *sequencer) ToggleStep(track, step int) {
 	s.tracks[track].steps[step].clearParameters()
 }
 
+// CopyStep copies a step to the 'clipboard' step.
+// TODO: Double check this, had some AI help with pointers
```
Comment (xaviergodart) [inline sequencer/sequencer.go]: Seems ok to me :ok_hand:

File: sequencer/sequencer.go [hunk 3]
```diff
@@ -245,6 +249,69 @@ func (s *sequencer) ToggleStep(track, step int) {
...
+func (s *sequencer) CopyStep(track, srcStep int) {
+	if track < 0 || track >= len(s.tracks) || srcStep < 0 || srcStep >= len(s.tracks[track].steps) {
+		return // Out of bounds, do nothing
+	}
+
+	originalStep := s.tracks[track].steps[srcStep]
+
+	// Create a deep copy of the step
+	s.stepClipboard = step{
+		midi:        originalStep.midi, // Assuming midi.Midi is safe to copy directly
```
Comment (xaviergodart) [inline sequencer/sequencer.go]: Yep, it's safe. There's only one midi instance and we're passing the pointer around where needed.

File: sequencer/sequencer.go [hunk 4]
```diff
@@ -280,3 +347,31 @@ func (s sequencer) sendControls() {
...
+func copyUint8SlicePtr(ptr *[]uint8) *[]uint8 {
+	if ptr == nil {
+		return nil
+	}
+	copy := make([]uint8, len(*ptr))
+	for i, v := range *ptr {
+		copy[i] = v
+	}
+	return &copy
+}
```
Comment (xaviergodart) [inline sequencer/sequencer.go]: Could you move these in a separate `helper.go` file? Let's keep it in the sequencer package for now.

File: ui/keymap.go [hunk 5]
```diff
@@ -40,6 +40,8 @@ type keyMap struct {
 
 	AddParam    key.Binding
 	RemoveParam key.Binding
+	CopyParams  key.Binding
+	PasteParams key.Binding
```
Comment (xaviergodart) [inline ui/keymap.go]: Same here, `CopyStep` and `PasteStep`

**Review by xaviergodart — CHANGES_REQUESTED** [review]
Works like a charm! Thanks!
It just needs some very minor things to tweak.

File: filesystem/keymap.go (hunk 1 above)
Comment (ajboni) [reply filesystem/keymap.go]: Makes sense, thanks!

File: sequencer/sequencer.go [hunk 6]
```diff
@@ -245,6 +249,68 @@ func (s *sequencer) ToggleStep(track, step int) {
 	s.tracks[track].steps[step].clearParameters()
 }
 
+// CopyStep copies a step to the 'clipboard' step.
+// TODO: Double check this, had some AI help with pointers
```
Comment (xaviergodart) [inline sequencer/sequencer.go]:
```suggestion
```

File: sequencer/sequencer.go [hunk 7]
```diff
@@ -245,6 +249,68 @@ func (s *sequencer) ToggleStep(track, step int) {
...
+func (s *sequencer) CopyStep(track, srcStep int) {
+	if track < 0 || track >= len(s.tracks) || srcStep < 0 || srcStep >= len(s.tracks[track].steps) {
+		return // Out of bounds, do nothing
+	}
+
+	originalStep := s.tracks[track].steps[srcStep]
+
+	// Create a deep copy of the step
+	s.stepClipboard = step{
+		midi:        originalStep.midi, // Assuming midi.Midi is safe to copy directly
```
Comment (xaviergodart) [inline sequencer/sequencer.go]:
```suggestion
		midi:        originalStep.midi,
```

**Review by xaviergodart — APPROVED** [review]
LGTM :ok_hand: Thanks!

**Pull request merged**